---
title: "Sandbox"
author: "Grace Saville"
date: "01/10/2021"
output: html_document
---
# Preamble 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r eval = FALSE}
# Installing phylogenetics packages
# install.packages('ctv')
# library('ctv')
# install.views('Phylogenetics')
# update.views('Phylogenetics')
library(plyr)
library(ape)
library(phylotools)
library(stringr)
library(dplyr)
```

## Looking at the data
```{r eval = FALSE}
data <- read.delim("Genotyping-007.010-01_copy.txt")
dim(data) #478 rows, 332 columns
data[1,1:16] # SNP data in columns 16 to 333
class(data[5,17]) # character
count(data$island) #  how many samples from each island there are
count(data[,206]) # how many of each base combination there are in a SNP column
```

## Making a subset
```{r eval = FALSE}
sub <- data[1:10,17:30] # making subset to experiment with
str(sub)
# sub[sub=="?"] <- NA # not sure if I want to replace the ?'s with NA's
summary(sub)
str(sub)

```

# Heterozygousity experimenting
```{r eval = FALSE}
x <- c("A:A", "G:T", "C:C", "A:T", "T:A","T:T", "A:C", "C:A") # test vector
hetz <- c("A:G", "G:A", "C:T", "T:C", "A:T", "T:A", "C:G", "G:C", "A:C", "C:A", "G:T", "T:G") # vector of heterozygous combinations
homz <- c("G:G", "A:A", "C:C", "T:T") # vector of homozygous combinations

# testing to see if I can identify what values are homozygous and heterozygous within an example vector
for (i in x){
  if (i %in% hetz){print("E")}
  else {print("O")}
}

# testing if I can replace the values in the above example vector 
for (i in 1:length(x)){
  if (x[i] %in% hetz){x[i] <- "E"}
  else {x[i] <- "O"}
}

# trying another method for the subset dataframe:
for (i in 1:nrow(sub)){
  sub[i,][sub[i,] %in% hetz] <- "E"
  sub[i,][sub[i,] %in% homz] <- "O"
}

# applying above for loop to entire df
z <- data # making a copy of the data
for (i in 1:nrow(z)){
  z[i,][z[i,] %in% hetz] <- "E"
  z[i,][z[i,] %in% homz] <- "O"
}

count(z[,105]) #column count
sum(str_count(z[35,], "E")) # row count

# could also try 0 for homo and 1 for hetero, therefore the lower the totals the lower the heterozygousity, however the number of NA's could interfere with the totals
z2 <- data # making a copy of the data
for (i in 1:nrow(z2)){
  z2[i,][z2[i,] %in% hetz] <- 1
  z2[i,][z2[i,] %in% homz] <- 0
}

z2[z2=="?"] <- NA # replacing ? with NA's 
for (i in 16:332) {class(z2[,i]) <- "numeric"} # changed the class to numeric for each column so the totals can be computed (wouldn't work without for loop for some reason?)

z2 <- rowwise(z2) # making the df rowwise so i can make calculations by row easier
names(z2)
z2 <- mutate(z2, total = sum(c_across(X11_CHR1_101004452:X154_CHR9_95290356), na.rm = TRUE))
z2[,333]

# trying this method of adding row totals at the end to the E/O df
z <- rowwise(z)
summarise(z)
mutate(z, Etotal = sum(str_count(z, "E")))
sum(str_count(z[35,], "?")) # row count

```

# Prep for making tree
```{r eval = FALSE}
dat2phylip(data, outfile = "test.phy") # testing this function that saves phylip files
```
Adapted IUPAC code:  

|Symbol|SNP bases|
|:----:|:----:|
|A|AA|
|T|TT|
|C|CC|
|G|GG|
|R (purine)|AG|
|r (purine)|GA|
|Y (pyrimidine)|CT|
|y (pyrimidine)|TC|
|W (weak)|AT|
|w (weak)|TA|
|S (strong)|CG|
|s (strong)|GC|
|M (amino)|AC|
|m (amino)|CA|
|K (keto)|GT|
|k (keto)|TG|

```{r}
y <- data # making a copy
# changing each SNP base combination to a single letter code
y[y=="?"] <- "-"
y[y=="A:A"] <- "A"
y[y=="T:T"] <- "T"
y[y=="C:C"] <- "C"
y[y=="G:G"] <- "G"
y[y=="A:G"] <- "R"
y[y=="G:A"] <- "r"
y[y=="C:T"] <- "Y"
y[y=="T:C"] <- "y"
y[y=="A:T"] <- "W"
y[y=="T:A"] <- "w"
y[y=="C:G"] <- "S"
y[y=="G:C"] <- "s"
y[y=="A:C"] <- "M"
y[y=="C:A"] <- "m"
y[y=="G:T"] <- "K"
y[y=="T:G"] <- "k"

names(y)
y <- y[,-c(2:15)] # removing the rows with information other than the species key and the code above
y <- tidyr::unite(y, bases, -1, sep = "", remove = TRUE) # merging all the base columns into one column so it can be converted easily into a phylip file,
y[,1] <- gsub(".", "", y[,1])
dat2phylip(y, outfile = "test.phy")
x <- read.phylip("test.phy")
```



# Mantel test prep
uses lat/long data, there are geospatial R packages that give exact distances on the globe between lat/long points which will be useful here 
```{r}

```
for later: HKY85 model for distances model
